### PR TITLE
Add support for message duplication check

### PR DESCRIPTION
**Pull request by yabinmeng**
It is invalid to detect message duplication when message de-duplication is enabled. Otherwise, it is possible to see duplicate messages.

### CODE REVIEW

File: driver-pulsar/src/main/java/io/nosqlbench/driver/pulsar/ops/PulsarConsumerOp.java [hunk 1]
```diff
@@ -136,22 +140,32 @@ public void run(Runnable timeTracker) {
...
-                        }
-                        // abnormal case: message loss
-                        else if ( (curMsgSeqId - prevMsgSeqid) > 1 ) {
-                            throw new RuntimeException("Detected message sequence id gap. Some published messages are not received!");
+                    if ( prevMsgSeqId > -1) {
+                        // normal case: message sequence id is monotonically increasing by 1
+                        if ((curMsgSeqId - prevMsgSeqId) != 1) {
+                            // abnormal case: out of ordering
+                            if (curMsgSeqId < prevMsgSeqId) {
+                                throw new RuntimeException("" +
```
Comment (jshook) [inline driver-pulsar/src/main/java/io/nosqlbench/driver/pulsar/ops/PulsarConsumerOp.java:152]: For each reason to throw an exception, can you create a specific exception type (that extends RuntimeException) and add it to the pulsar driver docs? This allows for configurable error handling.

**Review by jshook — CHANGES_REQUESTED** [review]
Can you add the specific exception type implementations?

File: driver-pulsar/src/main/java/io/nosqlbench/driver/pulsar/ops/PulsarConsumerOp.java (hunk 1 above)
Comment (yabinmeng) [reply driver-pulsar/src/main/java/io/nosqlbench/driver/pulsar/ops/PulsarConsumerOp.java:152]: Sure, I will add that

File: driver-pulsar/src/main/java/io/nosqlbench/driver/pulsar/exception/PulsarDriverUnexpectedException.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package io.nosqlbench.driver.pulsar.exception;
+
+public class PulsarDriverUnexpectedException extends RuntimeException {
+
```
Comment (jshook) [inline driver-pulsar/src/main/java/io/nosqlbench/driver/pulsar/exception/PulsarDriverUnexpectedException.java:4]: Unexpected what?

Comment (yabinmeng) [reply driver-pulsar/src/main/java/io/nosqlbench/driver/pulsar/exception/PulsarDriverUnexpectedException.java:4]: That's generic catch-all unexpected.

Comment (yabinmeng) [reply driver-pulsar/src/main/java/io/nosqlbench/driver/pulsar/exception/PulsarDriverUnexpectedException.java:4]: The added functionality has some urgency. If possible, please merge these changes as soon as we can. I can change names if needed later.

File: driver-pulsar/src/main/java/io/nosqlbench/driver/pulsar/ops/PulsarConsumerOp.java (hunk 1 above)
Comment (yabinmeng) [reply driver-pulsar/src/main/java/io/nosqlbench/driver/pulsar/ops/PulsarConsumerOp.java]: Added via commit 949790b797d2d24f22c9f02e7c1a7f013718357c

**Review by jshook — APPROVED** [review]
Merging this because Yabin requests urgency. Not the best reason to merge any change. Let's discuss this further.

**Pull request merged**